framework/src/meshmodifiers/BreakBoundaryOnSubdomain.C

```diff
@@ -0,0 +1,107 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "BreakBoundaryOnSubdomain.h"
+
+#include "MooseMesh.h"
+#include "MooseUtils.h"
+
+template<>
+InputParameters validParams<BreakBoundaryOnSubdomain>()
+{
+  InputParameters params = validParams<MeshModifier>();
+  params.addClassDescription("Break boundaries based on the subdomains to which their sides are attached. Naming convention for the new boundaries will be the old boundary name plus \"_to_\" plus the subdomain name");
+  params.addParam<std::vector<BoundaryName> >("boundaries", "Boundaries to be broken. Default means to break all boundaries");
+  return params;
+}
+
+BreakBoundaryOnSubdomain::BreakBoundaryOnSubdomain(const InputParameters & parameters) :
+    MeshModifier(parameters)
+{
+}
+
+void
+BreakBoundaryOnSubdomain::modify()
+{
+  // get the mesh and boundary info
+  MeshBase & mesh = _mesh_ptr->getMesh();
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+
+  // get IDs of all boundaries to be broken
+  std::set<BoundaryID> breaking_boundary_ids;
+  if (isParamValid("boundaries"))
+  {
+    std::vector<BoundaryName> boundary_names = getParam<std::vector<BoundaryName> >("boundaries");
+    for (unsigned int i = 0; i < boundary_names.size(); ++i)
+      breaking_boundary_ids.insert(_mesh_ptr->getBoundaryID(boundary_names[i]));
+  }
+  else
+    breaking_boundary_ids = boundary_info.get_boundary_ids();
+
+  const MeshBase::const_element_iterator end_el = mesh.active_elements_end();
+
+  // create a list of new boundary names
+  std::set<BoundaryName> new_boundary_name_set;
+  for (MeshBase::const_element_iterator el = mesh.active_elements_begin(); el != end_el; ++el)
+  {
+    Elem * elem = *el;
+    unsigned int subdomain_id = elem->subdomain_id();
```

Comment:
I disagree - people are often, and perhaps even "usually" wrong about types. Every single time you have written `for (unsigned int i = 0; i < vec.size(); ++i)` you used the wrong type. Most variables can be disambiguated just by using better names.